### PR TITLE
sd-dhcp-server: clear out expired leases when finding new address

### PR DESCRIPTION
**Pull request by tohojo**
The DHCP server configuration supports setting a maximum lease time, but old
leases are never actually cleared out if the client doesn't send a RELEASE.
This causes the pool to run out of addresses on networks where clients just
disappear, which is a fairly common occurrence on wireless networks.

Fix this by checking the expiry time of old leases when trying to find an
address to hand out, and clearing out stale entries so the address can be
reused for new clients.

While we're add it, add a couple of checks to make sure to never hand out the
server's own address (neither in response to DISCOVER or REQUEST messages).

To see the failure, run systemd-networkd with a .network file like this:

```
[Match]
Name=veth0

[Network]
DHCPServer=yes
Address=192.168.0.1/30
IPv6AcceptRA=no

[DHCPServer]
MaxLeaseTimeSec=20s
```

This should create a pool with only a single address in it. Running three DHCP
client requests like:

``` shell
$ dhcpcd -41BAG --noconfigure -I 01:02:01 veth1
$ dhcpcd -41BAG --noconfigure -I 01:02:02 veth1
$ dhcpcd -41BAG --noconfigure -I 01:02:03 veth1
```

will result in two offers of 192.168.0.1 and 192.168.0.2, but the third request
will fail indefinitely.

After the patch, the first request will succeed (returning 192.168.0.2), and
the second will initially fail, but succeed when the original entry expires
after 20 seconds. 192.168.0.1 will never be handed out after applying this
patch.

### CODE REVIEW

File: src/libsystemd-network/sd-dhcp-server.c [hunk 1]
```diff
@@ -854,6 +854,31 @@ static int prepare_new_lease(
         return 0;
 }
 
+static bool address_is_available(sd_dhcp_server *server, be32_t address) {
+        DHCPLease *lease;
+        usec_t time_now;
+        int r;
+
```
Comment (bluca) [inline src/libsystemd-network/sd-dhcp-server.c:861]: `assert(server)`

Comment (tohojo) [reply src/libsystemd-network/sd-dhcp-server.c:861]: Done :)

File: src/libsystemd-network/sd-dhcp-server.c [hunk 2]
```diff
@@ -854,6 +854,33 @@ static int prepare_new_lease(
...
+                return false; /* address in static lease, or server addr so unavailable */
+
+        lease = hashmap_get(server->bound_leases_by_address, UINT32_TO_PTR(address));
+        if (!lease)
+                return true; /* address not bound, so available */
+
+        /* address in existing lease, check expiry */
+        r = sd_event_now(server->event, clock_boottime_or_monotonic(), &time_now);
+        if (r < 0)
+                return false;
```
Comment (bluca) [inline src/libsystemd-network/sd-dhcp-server.c]: leaks the lease? You probably want to declare as: `_cleanup_(dhcp_lease_freep) DHCPLease *lease = NULL;`

Comment (yuwata) [reply src/libsystemd-network/sd-dhcp-server.c]: This should be critical. Return `r`.

File: src/libsystemd-network/sd-dhcp-server.c [hunk 3]
```diff
@@ -1040,7 +1066,8 @@ int dhcp_server_handle_message(sd_dhcp_server *server, DHCPMessage *message, siz
 
                         return DHCP_ACK;
 
-                } else if (pool_offset >= 0 && existing_lease_by_address == existing_lease) {
+                } else if (pool_offset >= 0 && existing_lease_by_address == existing_lease &&
+                           address != server->address) {
```
Comment (yuwata) [inline src/libsystemd-network/sd-dhcp-server.c]: This new condition must be earlier. Could you move just before `pool_offset = get_pool_offset(server, address);` in the above?

Also, could you split this change into another commit?

File: src/libsystemd-network/sd-dhcp-server.c [hunk 4]
```diff
@@ -854,6 +854,33 @@ static int prepare_new_lease(
         return 0;
 }
 
+static bool address_is_available(sd_dhcp_server *server, be32_t address) {
```
Comment (yuwata) [inline src/libsystemd-network/sd-dhcp-server.c]: The expired lease should be cleaned up just after the following
```
        r = ensure_sane_request(server, req, message);
        if (r < 0)
                /* this only fails on critical errors */
                return r;
```

So, could you split this function into two? Say, `dhcp_server_cleanup_expired_leases()` and `address_is_available()`. Then, call `dhcp_server_cleanup_expired_leases()` after `ensure_sane_request()`?

File: src/libsystemd-network/sd-dhcp-server.c [hunk 5]
```diff
@@ -854,6 +854,42 @@ static int prepare_new_lease(
...
+                be32_t address;
+
+                address = server->subnet | htobe32(server->pool_offset + i);
+                lease = hashmap_get(server->bound_leases_by_address, UINT32_TO_PTR(address));
+
+                if (lease && lease->expiration < time_now) {
+                        log_dhcp_server(server, "CLEAN (0x%x)", be32toh(address));
+                        dhcp_lease_free(lease);
+                }
+        }
```
Comment (yuwata) [inline src/libsystemd-network/sd-dhcp-server.c]: Let's simplify this more like below:
```
HASHMAP_FOREACH(lease, server->bound_leases_by_client_id)
        if (lease->expiration < time_now) {
                log_dhcp_server(server, "CLEAN (0x%x)", lease->address);
                dhcp_lease_free(lease);
        }
```

Comment (tohojo) [reply src/libsystemd-network/sd-dhcp-server.c]: Ah yes, that's better!
